lib/evaluate/fold.cc

```diff
@@ -152,35 +157,486 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
+  if constexpr (TOCAT == TypeCategory::Integer || TOCAT == TypeCategory::Real) {
+    return FROMCAT == TypeCategory::Real || FROMCAT == TypeCategory::Integer;
+  } else if constexpr (TOCAT == TypeCategory::Logical) {
+    return FROMCAT == TypeCategory::Logical;
+  }
+  return false;
+}
+
+template<typename TO, typename Operand>
+Scalar<TO> ScalarConvert(
+    FoldingContext &context, const Scalar<Operand> &value) {
+  static_assert(IsAllowedBasicConversion<TO::category, Operand::category>());
+  if constexpr (TO::category == TypeCategory::Integer) {
+    if constexpr (Operand::category == TypeCategory::Integer) {
+      auto converted{Scalar<TO>::ConvertSigned(value)};
+      if (converted.overflow) {
+        context.messages.Say(
+            "INTEGER(%d) to INTEGER(%d) conversion overflowed"_en_US,
+            Operand::kind, TO::kind);
+      }
+      return std::move(converted.value);
+    } else if constexpr (Operand::category == TypeCategory::Real) {
+      auto converted{value.template ToInteger<Scalar<TO>>()};
+      if (converted.flags.test(RealFlag::InvalidArgument)) {
+        context.messages.Say(
+            "REAL(%d) to INTEGER(%d) conversion: invalid argument"_en_US,
+            Operand::kind, TO::kind);
+      } else if (converted.flags.test(RealFlag::Overflow)) {
+        context.messages.Say(
+            "REAL(%d) to INTEGER(%d) conversion overflowed"_en_US,
+            Operand::kind, TO::kind);
+      }
+      return std::move(converted.value);
+    }
+  } else if constexpr (TO::category == TypeCategory::Real) {
+    char buffer[64];
+    if constexpr (Operand::category == TypeCategory::Integer) {
+      auto converted{Scalar<TO>::FromInteger(value)};
+      if (!converted.flags.empty()) {
+        std::snprintf(buffer, sizeof buffer,
+            "INTEGER(%d) to REAL(%d) conversion", Operand::kind, TO::kind);
+        RealFlagWarnings(context, converted.flags, buffer);
+      }
+      return std::move(converted.value);
+    } else if constexpr (Operand::category == TypeCategory::Real) {
+      auto converted{Scalar<TO>::Convert(value)};
+      if (!converted.flags.empty()) {
+        std::snprintf(buffer, sizeof buffer, "REAL(%d) to REAL(%d) conversion",
+            Operand::kind, TO::kind);
+        RealFlagWarnings(context, converted.flags, buffer);
+      }
+      if (context.flushSubnormalsToZero) {
+        converted.value = converted.value.FlushSubnormalToZero();
+      }
+      return std::move(converted.value);
+    }
+  } else if constexpr (TO::category == TypeCategory::Logical &&
+      Operand::category == TypeCategory::Logical) {
+    return value.IsTrue();
+  }
+}
+
+// helpers to fold intrinsic function references
+namespace intrinsicHelper {
+// helper to produce hash of intrinsic names based the first 3 letters. All
+// intrinsic names are longer than 3 letters
+static constexpr inline std::int32_t CommonHash(const char *s, std::size_t n) {
+  if (n < 3) {
```

Comment:
Shouldn't this be a static_assert? It seems like the intention here is that all strings will be at least 3 chars long, per the comment.

Reply:
When CommonHash is used in the _hash string literal suffix it could be a static assert, but when called from DynamicHash, it cannot be a static assert anymore so I went for this. However, maybe dying is a better alternative than silently returning 0, I'll wait for Peter answer to the other comment regarding hashing names and add this if it remains.

Reply:
I think the `constexpr` had me thinking this function was used at compile-time only.